src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/Processors/LedgerTransactionMarkers/ManifestProcessor.cs

```diff
@@ -108,128 +108,169 @@ public void OnTransactionScan(CoreModel.CommittedTransaction transaction, long s
+                        Id = _context.Sequences.LedgerTransactionMarkerSequence++,
+                        StateVersion = stateVersion,
+                        OperationType = LedgerTransactionMarkerOperationType.BadgePresented,
+                        EntityId = referencedEntity.DatabaseId,
+                    });
             }
 
-            foreach (var address in extractedAddresses.ResourceAddresses)
+            foreach (var resourceAddress in extractedAddresses.ResourceAddresses)
             {
-                if (_referencedEntities.TryGet(address, out var re))
+                if (!_referencedEntities.TryGet(resourceAddress, out var referencedEntity))
                 {
-                    _ledgerTransactionMarkersToAdd.Add(
-                        new ManifestAddressLedgerTransactionMarker
-                        {
-                            Id = _context.Sequences.LedgerTransactionMarkerSequence++,
-                            StateVersion = stateVersion,
-                            OperationType = LedgerTransactionMarkerOperationType.ResourceInUse,
-                            EntityId = re.DatabaseId,
-                        });
+                    throw new UnreachableException($"Entity: {resourceAddress} was not present in referenced entities dictionary.");
                 }
+
+                ledgerTransactionMarkersToAdd.Add(
+                    new ManifestAddressLedgerTransactionMarker
+                    {
+                        Id = _context.Sequences.LedgerTransactionMarkerSequence++,
+                        StateVersion = stateVersion,
+                        OperationType = LedgerTransactionMarkerOperationType.ResourceInUse,
+                        EntityId = referencedEntity.DatabaseId,
+                    });
             }
 
-            foreach (var address in extractedAddresses.AccountsRequiringAuth)
+            foreach (var addressWithEntityType in extractedAddresses.AccountsRequiringAuth)
             {
-                if (_referencedEntities.TryGet(address, out var re))
+                if (_referencedEntities.TryGet(addressWithEntityType.Address, out var referencedEntity))
                 {
-                    _ledgerTransactionMarkersToAdd.Add(
+                    ledgerTransactionMarkersToAdd.Add(
                         new ManifestAddressLedgerTransactionMarker
                         {
                             Id = _context.Sequences.LedgerTransactionMarkerSequence++,
                             StateVersion = stateVersion,
                             OperationType = LedgerTransactionMarkerOperationType.AccountOwnerMethodCall,
-                            EntityId = re.DatabaseId,
+                            EntityId = referencedEntity.DatabaseId,
                         });
                 }
+                else if (addressWithEntityType.EntityType is not (ToolkitModel.EntityType.GlobalVirtualEd25519Account or ToolkitModel.EntityType.GlobalVirtualSecp256k1Account))
```

Comment:
This is incorrect and flakey (e.g. it misses virtual identities).

Could we instead have a `is_preallocated()` method on the address type, which delegates to an `is_preallocated()` method on the entity type?